en/lessons/libraries/stream-data.md

```diff
@@ -0,0 +1,397 @@
+
+The StreamData library will run each test 100 times with random data. When a test fails, StreamData will try to shrink the input to the smallest value that causes the test failure. This can be helpful when you have to debug your code! If a 50-element list causes your function to break, and only one of the list elements is problematic, StreamData can help you identify the offending element.
+
+This testing library has two main modules. [`StreamData`](https://hexdocs.pm/stream_data/StreamData.html) generates streams of random data. [`ExUnitProperties`](https://hexdocs.pm/stream_data/ExUnitProperties.htm) lets you run tests against your functions, using the generated data as your input.
+
+You might be asking how you can say anything meaningful about a function if you don't know what your exact inputs are. Read on!
+
+
+## Using StreamData
+
+To illustrate the features of StreamData, we'll write a few simple utility functions that repeat values. Let's say we want a function like [`String.duplicate/2`](https://hexdocs.pm/elixir/String.html#duplicate/2), but one that will duplicate strings, lists, or tuples.
+
+
+### Strings
+
+Let's duplicate strings first. What are some requirements for our function?
+
+1. The first argument should be a string. This is the string that we'll duplicate.
+1. The second argument should be a non-negative integer. It shows how many times we'll duplicate the first argument.
+1. The function should return a string. This string zero or more of the is the first string, repeated over and over.
+
+When we run our function, we want it to look like this:
+
+```elixir
+Repeater.duplicate("a", 4)
+# "aaaa"
+```
+
+Elixir has a function, `String.duplicate/2` that will handle this for us. Our new `duplicate/2` will just delegate to that function:
+
+```elixir
+defmodule Repeater do
+  def duplicate(string, times) when is_binary(string) do
+    String.duplicate(string, times)
+  end
+end
+```
+
+The happy path should be easy to test with [ExUnit](https://hexdocs.pm/ex_unit/ExUnit.html).
+
+
+```elixir
+defmodule RepeaterTest do
+  use ExUnit.Case
+
+  describe "duplicate/2" do
+    test "creates a new string, with the first argument duplicated a specified number of times" do
+      assert "aaaa" == Repeater.duplicate("a", 4)
+    end
+  end
+end
+```
+
+That's hardly a comprehensive test, though. What should happen when the second argument is `0`? What should the output be when the first argument is an empty string? What does it even mean to repeat an empty string? How should the function work with UTF-8 characters? Will the function still work with large input strings?
+
+You might say that my criticisms are unfair. After all, who would write just one example to cover this function? I'll grant your point. We could write more examples to test edge cases and large strings. However, let's see if we can use StreamData to test this function more rigorously without much more code.
```

Comment:
Since this is not a blog post but a lesson, it might be better to remove the discussion part of this paragraph.

Reply:
Thanks for the suggestion :+1: Would removing all sentences except the last two be better? I'll push up that change, and you can let me know what you think.

Reply:
I would remove this part: `You might say that my criticisms are unfair. After all, who would write just one example to cover this function? I'll grant your point` and it's back to a lesson format :)